### PR TITLE
feat: NanoTDF Implementation

### PR DESCRIPTION
**Pull request by mustyantsev**

### CODE REVIEW

File: sdk/pom.xml [hunk 1]
```diff
@@ -10,8 +10,8 @@
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-compiler-plugin</artifactId>
                 <configuration>
-                    <source>11</source>
-                    <target>11</target>
+                    <source>15</source>
```
Comment (ttschampel) [inline sdk/pom.xml:13]: I think we want to keep this at 11 if possible until some downstream projects can upgrade (gateway)

**Review by ttschampel — COMMENTED** [review]
Can we add a section in the README about using the Nano Encrypt feature?

File: sdk/src/main/java/io/opentdf/platform/sdk/Config.java [hunk 2]
```diff
@@ -79,4 +83,63 @@ public static Consumer<TDFConfig> withMetaData(String metaData) {
...
+        };
+    }
+
+    public static Consumer<NanoTDFConfig> withEllipticCurve(String curve) {
+        NanoTDFType.ECCurve ecCurve;
+        if (curve.compareToIgnoreCase(NanoTDFType.ECCurve.SECP384R1.toString()) == 0) {
+            ecCurve = NanoTDFType.ECCurve.SECP384R1;
+        } else if (curve.compareToIgnoreCase(NanoTDFType.ECCurve.SECP521R1.toString()) == 0) {
+            ecCurve = NanoTDFType.ECCurve.SECP521R1;
+        } else {
```
Comment (patmantru) [inline sdk/src/main/java/io/opentdf/platform/sdk/Config.java:136]: should have at least a warning here that the supplied curve string was not recognized before using the default.

File: sdk/src/main/java/io/opentdf/platform/sdk/NanoTDF.java [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+        return nanoTDFSize;
+    }
+
+    public void readNanoTDF(ByteBuffer nanoTDF, OutputStream outputStream,
+                            SDK.KAS ka) {
+
+
+
+
+        /*
```
Comment (patmantru) [inline sdk/src/main/java/io/opentdf/platform/sdk/NanoTDF.java:161]: lol...looks familiar.  :)

File: sdk/src/main/java/io/opentdf/platform/sdk/nanotdf/ECKeyPair.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    }
+    private KeyPair keyPair;
+    private String curveName;
+
+    public ECKeyPair() throws NoSuchAlgorithmException, NoSuchProviderException, InvalidAlgorithmParameterException {
+        this("secp256r1", ECAlgorithm.ECDH);
+    }
+
+    public ECKeyPair(String curveName, ECAlgorithm algorithm) throws NoSuchAlgorithmException, InvalidAlgorithmParameterException,
+            NoSuchProviderException {
```
Comment (patmantru) [inline sdk/src/main/java/io/opentdf/platform/sdk/nanotdf/ECKeyPair.java:64]: the signature says it throws those exceptions, but I don't see where they would be coming from.  It *should be* checking parms and throwing those as appropriate though.

File: sdk/src/main/java/io/opentdf/platform/sdk/nanotdf/NanoTDFType.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+public class NanoTDFType {
+    public enum ECCurve {
+        SECP256R1("secp256r1"),
+        SECP384R1("secp384r1"),
+        SECP521R1("secp384r1"),
+        SECP256K1("secp256k1");
+
+        private String name;
+
+        ECCurve(String curveName) {
```
Comment (patmantru) [inline sdk/src/main/java/io/opentdf/platform/sdk/nanotdf/NanoTDFType.java:12]: Should have argument validation here - throw exception if the string does not match one of the enum values

**Review by patmantru — CHANGES_REQUESTED** [review]

**Review by patmantru — COMMENTED** [review]
Changing to comment-only I think?

**Comment by ttschampel** [comment]
Any idea why the checks aren't running for this PR?

File: sdk/pom.xml (hunk 1 above)
Comment (ttschampel) [reply sdk/pom.xml:13]: The build is breaking b/c of this

Comment (sujankota) [reply sdk/pom.xml:13]: will fix it.

File: sdk/src/main/java/io/opentdf/platform/sdk/Config.java (hunk 2 above)
Comment (biscoe916) [reply sdk/src/main/java/io/opentdf/platform/sdk/Config.java:136]: I agree. If no curve is provided, we should default, but if an invalid curve was supplied we should fail.

**Review by patmantru — DISMISSED** [review]

**Review by patmantru — APPROVED** [review]

**Pull request merged**